### ISSUE TITLE
Usage Questions: Transactional Commands + Edit

### ISSUE THREAD

**Issue by thewatts** [rootstrap/yaaf#66]
Hello again! 👋 

As we're evaluating YAAF - we're looking at how we would go about adjusting current implementations to leverage what YAAF provides.

We have a couple of specific scenarios:

#### 1. Transactional Commands
In our current form implementations, we have (often order-dependent) methods that need to be called _within_ the `save` transaction. Should they fail - we want the model persistence to rollback.

Ex:

```ruby
def save
  if valid?
    ApplicationRecord.transaction do
      end_current_session # updates an existing Session record 
      remove_clients_from_other_sessions # calls out to Form that raises on error
      create_new_session # create a new Session record
      create_missing_documentation # creates Documentation records
      migrate_shared_data # updates other records
    end
  end
end
```

My only thought here is that - in order for this to work - each of those methods would have to delegate out to an object that in itself would live within the `@models` array, to be saved (?)

#### 2. Editing
I feel like using form objects on an edit page isn't really talked about a lot in the examples given throughout the community.

In the case of editing, we keep having to lean on overriding the `reader` method for attributes so that we can _either_ pull them from the params given, _or_ from the record that was already saved.

> Note: This is a trivial example, but highlights the use case of sharing the same form for `new` and `edit`

```erb
<%= form_for @registration do |form| %>
  <%= form.text_field :user_email %>
<% end %>
```

```ruby
class EventRegistration
  include ActiveModel::Model

  attr_accessor :id, :event_id, :user_email

  validates :event_id, presence: true
  validates :user_email, presence: true

  def registration
    @registration ||= Registration.find_by(id: id) || Registration.new
  end

  def persisted?
    registration.persisted?
  end

  def user_email
    super.presence || registration.user_email
  end
end
```

---

Thanks for your time / help!

**Comment by thewatts** [comment]
🤦‍♂️ so, in digging deeper into what's going on - it looks like I could (for #1), use `after_save` callbacks, as they are taking place within the transaction block.

Sorry for the poor question!

Looking forward to your thoughts on #2 😄

Thanks!

**Comment by santib** [comment]
Hey @thewatts 👋 Regarding `#2` I agree with what you said. My opinion is that if the logic differs between the `new` and `edit` actions, then it's better to have completely separate form objects (e.g. `NewHouseForm` and `EditHouseForm`). But in general, that hasn't been my experience in real-world apps, and instead, we share the same form object for the `new` and `edit` actions (e.g. `HouseForm`), and as you said we have a method that either finds or initialize the record. For example

```ruby
class HouseForm < YAAF::Form
  attr_accessor :id, :address

  def initialize(attributes)
    super(attributes)
    @models = [house]
  end

  def house
    @house ||= House.find_or_initialize_by(id: id).tap do |house|
      house.address = address
    end
  end
end
```

If there are things that you only want to do on initialization you can pass a block to `find_or_initialize_by` as well.

**Comment by thewatts** [comment]
Interesting! Thanks for the quick response, @santib!

Given how typical Rails `update` actions work, along with strong params, in the above example would you:

```ruby
# Instantiate the form with the id, and then separately assign the strong param attributes
def update
  @house_form = HouseForm.new(id: params[:id])
  @house_form.assign_attributes(house_params)

  if @house_form.save
    redirect_to @house_form,
      notice: "Yay!"
  else
    render :edit
  end
end

private

def house_params
  params
    .require(:house)
    .permit(:address)
end
``` 

or

```ruby
# inject the `id` into the house_params
def update
  @house_form = HouseForm.new(house_params)

  if @house_form.save
    redirect_to @house_form,
      notice: "Yay!"
  else
    render :edit
  end
end

private

def house_params
  params
    .require(:house)
    .permit(:id, :address)
    .tap { |attributes| attributes[:id] = params[:id] }
end
```

--

A bit of a nuanced question, I know - but we're curious how other teams have been solving these problems.

( Thank you! )

**Comment by santib** [comment]
@thewatts I think I'd go with this

```ruby
def update
  @house_form = HouseForm.new(house_params.merge(id: params[:id]))

  if @house_form.save
    redirect_to @house_form, notice: "Yay!"
  else
    render :edit
  end
end

private

def house_params
  params.require(:house).permit(:address)
end
```

so that we don't alter the strong params and we don't assign attributes from outside. But it's just a matter of preference I guess, and all of them would be fine

**Comment by thewatts** [comment]
@santib thanks for the feedback and thoughts!

Much appreciated!

**Comment by thewatts** [comment]
Hey @santib - 

> we share the same form object for the new and edit actions 

I've been trying to replicate this, given the example above, and I've been having a bit of trouble. Would love some insight into how you're all solving this problem.

My main issue - the form helpers in the view aren't able to pull values in the `edit` scenario, as those methods on the `HouseForm` return what it's received in the `params`.

As such - the edit form is blank:

![CleanShot 2021-07-15 at 14 25 43](https://user-images.githubusercontent.com/1141717/125845911-d1954384-77a3-4f45-9918-dd2ba0dd8500.png)

Here's a sample project with the code: https://github.com/thewatts/form_samples

Thanks!

**Comment by thewatts** [comment]
@santib I took a stab at how I would solve this - but it seems pretty heavy-handed.

Maybe that's just par for the course given the abstraction that's going on and how Rails helpers work, but would love to know if you all have an alternate/better way.

https://github.com/thewatts/form_samples/pull/1/commits/994b1f0b83fe08abc1758baa7cc6d33677451e66

**Comment by santib** [comment]
Hey @thewatts, good point I think we should provide more help in the README regarding this 😅 

What I usually do is using the `super` call with the corresponding attributes. Currently, in the sample app, you have

```ruby
super(attributes)
```
that is perfect, but for the `edit` action you'd need something like

```ruby
super(house.slice(:name, :address))
```

If you separate the form objects then that should be enough. If you want to have one form object for both `new` and `edit` actions then it'd probably look like this
```ruby
attributes = Hash(house&.slice(:name, :address)).merge(attributes)
super(attributes)
```

so that permitted parameters override the existing ones (specifically thinking on the case of an error during the `edit` action)

But yeah I guess this is how I would do it, but there could be many other ways equally valid. Thoughts?

cc'ing @juanmanuelramallo in case he wants to share his experience

**Comment by thewatts** [comment]
```ruby
super(Hash(house&.slice(:name, :address)).merge(attributes))
```

Hmm - would this work in the case of a single form object?

`house` is initialized by the `id` and then attributes are (re)assigned in the block - wouldn't the above still give us empty attribute values?

```ruby
def house
  @house ||= House.find_or_initialize_by(id: id).tap do |house|
    house.name = name
    house.address = address
  end
end
```

It would work if we _didn't_ do the `.tap {}` assignment process, but then we'd have to assign the attributes at some point before saving.

---

Also - we both responded around the same time - just confirming you saw: https://github.com/rootstrap/yaaf/issues/66#issuecomment-880964731

**Comment by santib** [comment]
@thewatts Great points there, I went ahead and tried myself how I would do it in your sample app and this is the result

![image](https://user-images.githubusercontent.com/6373536/125852378-9471c890-2495-42d3-be41-04cdb4375016.png)


As you correctly mentioned, in my previous comment there was a problem with the `id`, and also the way you were assigning attributes wasn't behaving in a good way. So the changes were basically having the new attributes accessible and use them directly in the `house` method. And then using the house with updated values for the `super` method. WDYT?

I saw https://github.com/rootstrap/yaaf/issues/66#issuecomment-880964731 but I agree that's too manual
